src/server/pps/server/extended_trace.go

```diff
@@ -0,0 +1,79 @@
+package server
+
+import (
+	"context"
+	"fmt"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	opentracing "github.com/opentracing/opentracing-go"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/tracing"
+	"github.com/pachyderm/pachyderm/src/client/pkg/tracing/extended"
+)
+
+func addPipelineSpanToAnyExtendedTrace(ctx context.Context, c *etcd.Client, pipeline, operation string, kvs ...interface{}) (opentracing.Span, context.Context) {
+	if !tracing.IsActive() {
+		return nil, ctx // no Jaeger instance to send trace info to
+	}
+
+	// use key from 'ctx' to read full ExtendedTrace reference from etcd
+	tracesCol := extended.TracesCol(c).ReadOnly(ctx)
+	var extendedTraceID string
+	var extendedTrace extended.TraceProto
+	var spanCtx opentracing.SpanContext
+	// options can't be nil, but there should only be at most one trace attached
+	// to 'pipeline' so the options shouldn't matter
+	if err := tracesCol.GetByIndex(extended.PipelineIndex, pipeline, &extendedTrace, extended.TraceGetOpts,
+		func(key string) error {
+			if spanCtx != nil {
+				return fmt.Errorf("second, unexpected span associated with pipeline %q: %s",
+					pipeline, key)
+			}
+			extendedTraceID = key
+
+			// Create new opentracing span from 'extendedTrace'
+			var err error
+			spanCtx, err = opentracing.GlobalTracer().Extract(opentracing.TextMap,
+				opentracing.TextMapCarrier(extendedTrace.SerializedTrace))
+			if err != nil {
+				return fmt.Errorf("could not extract span context from ExtendedTrace proto: %v", err)
+			}
+			return nil
+		}); err != nil {
+		log.Errorf("tracing error: %v", err)
+		return nil, ctx
+	}
+	if spanCtx == nil {
+		return nil, ctx
+	}
+
+	// TODO(msteffen) we may leak a trace if a user updates a pipeline with
+	// reprocess=false (so worker never restarts) AND the pipeline has no output
```

Comment:
Even with `reprocess=false` we still restart the workers, because you can still change the image and configuration of the pipeline. There's presently no way to update a pipeline without recreating workers although that's been requested so we may add that at some point. But for now I think this is a non-issue.

Reply:
Awesome—comment removed